### PR TITLE
Implement zoneinfo64 rules

### PR DESCRIPTION
**Pull request by robertbastian**
#6792

### CODE REVIEW

File: utils/calendrical_calculations/src/iso.rs [hunk 1]
```diff
@@ -14,8 +14,15 @@ use crate::rata_die::RataDie;
 const EPOCH: RataDie = RataDie::new(1);
 
 /// Whether or not `year` is a leap year
+///
+/// Inspired by Neri-Schneider <https://arxiv.org/abs/2102.06959>
 pub const fn is_leap_year(year: i32) -> bool {
-    year % 4 == 0 && (year % 400 == 0 || year % 100 != 0)
+    // This is branch-free, as it compiles to a conditional move
+    if year % 100 != 0 {
```
Comment (cassioneri) [inline utils/calendrical_calculations/src/iso.rs]:
```suggestion
    if year % 25 != 0 {
```
This change is correct and saves one `ror` instruction. (See https://godbolt.org/z/8KGoMGoTb)

File: utils/calendrical_calculations/src/iso.rs [hunk 2]
```diff
@@ -25,53 +32,62 @@ pub const fn is_leap_year(year: i32) -> bool {
...
-    };
+    fixed += days_before_month(year, month) as i64;
     // Days passed in current month
     fixed += day as i64;
     RataDie::new(fixed)
 }
 
+/// The number of days in this year before this month starts
+///
+/// Inspired by Neri-Schneider <https://arxiv.org/abs/2102.06959>
```
Comment (cassioneri) [inline utils/calendrical_calculations/src/iso.rs]:
```suggestion
/// Inspired by Neri-Schneider <https://onlinelibrary.wiley.com/doi/10.1002/spe.3172>
```
Please refer to the peer-reviewed and published version of the paper. It's much clearer than the preprint.

File: utils/calendrical_calculations/src/iso.rs [hunk 3]
```diff
@@ -14,8 +14,15 @@ use crate::rata_die::RataDie;
 const EPOCH: RataDie = RataDie::new(1);
 
 /// Whether or not `year` is a leap year
+///
+/// Inspired by Neri-Schneider <https://arxiv.org/abs/2102.06959>
```
Comment (cassioneri) [inline utils/calendrical_calculations/src/iso.rs]:
```suggestion
/// Inspired by Neri-Schneider <https://www.youtube.com/watch?v=J9KijLyP-yg&t=1239s>
```
This algorithm is presented in my talk but not in the cited paper.

**Comment by Manishearth** [comment]
@cassioneri n.b. the changes you are reviewing are from https://github.com/unicode-org/icu4x/pull/6870 , this PR is based on that. That's my bad, I accidentally linked to this PR instead of #6870 in the other issue.

Also, this specific PR is probably not going to be merged, it was presented as an alternate approach to https://github.com/unicode-org/icu4x/pull/6842 since @robertbastian  didn't like some of the changes there, but we've talked and since then I've split out some of the changes in https://github.com/unicode-org/icu4x/pull/6872 and will update #6842 to be based on that later. My new PR will likely use a bunch of the new algorithms from your paper added in #6870.

File: utils/calendrical_calculations/src/iso.rs (hunk 3 above)
Comment (cassioneri) [reply utils/calendrical_calculations/src/iso.rs]: Suggestion: instead of this
```
fn days_in_provided_month(year: i32, month: u8) -> u8 {
    match month {
        4 | 6 | 9 | 11 => 30,
        2 => 28 + is_leap_year2(year) as u8,
        1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
        _ => 0,
    }
}
```
use this
```
fn days_in_provided_month2(year: i32, month: u8) -> u8 {
    if month == 2 {
        28 + is_leap_year2(year) as u8
    } else {
        30 | month ^ (month >> 3)
    }
}
```
https://www.youtube.com/watch?v=J9KijLyP-yg&t=1394s
https://godbolt.org/z/GP51KeGs9

File: utils/calendrical_calculations/src/chinese_based.rs [hunk 4]
```diff
@@ -33,11 +33,23 @@ pub trait ChineseBased {
 
 /// Given an ISO year, return the extended year
 pub fn extended_from_iso<C: ChineseBased>(iso_year: i32) -> i32 {
-    iso_year - const { crate::iso::iso_year_from_fixed(C::EPOCH) as i32 - 1 }
+    iso_year
+        - const {
+            let Ok(y) = crate::iso::iso_year_from_fixed(C::EPOCH) else {
+                panic!()
```
Comment (Manishearth) [inline utils/calendrical_calculations/src/chinese_based.rs:39]: issue: we should not be panicking

File: utils/zoneinfo64/src/rule.rs [hunk 5]
```diff
@@ -203,42 +318,214 @@ impl Rule<'_> {
...
     /// Returns None if the seconds_since_epoch is not in range of the Rule.
     ///
     /// seconds_since_epoch must resolve to a year that is in-range for i32
-    pub(crate) fn resolve_utc(&self, _seconds_since_epoch: i64) -> Option<Offset> {
-        // Unimplemented
-        None
+    pub(crate) fn resolve_utc(&self, seconds_since_epoch: i64) -> Option<Offset> {
+        let year =
+            iso::iso_year_from_fixed(super::EPOCH + (seconds_since_epoch / SECONDS_IN_UTC_DAY))
+                .unwrap();
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs]: issue: no panics, please.

File: utils/zoneinfo64/src/rule.rs [hunk 6]
```diff
@@ -126,22 +130,26 @@ impl TzRule {
             .unwrap(),
         }
     }
+
+    fn is_inverted(&self) -> bool {
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs]: issue: I tend to find names like "is_inverted" to be very confusing, since it's not always clear what "regular" is. It would be nice if this were descriptive

that said, "inverted" appears to be a term of art here, so maybe it's fine. would like docs then

File: utils/zoneinfo64/src/rule.rs [hunk 7]
```diff
@@ -203,42 +318,214 @@ impl Rule<'_> {
...
+            minute,
+            second,
+        };
+
+        struct LazyRuleEval<'a> {
+            year: i32,
+            rule: &'a TzRuleDate,
+            standard_offset_seconds: i32,
+            additional_offset_seconds: i32,
+            correction: i32,
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs]: question: what's this?

File: utils/zoneinfo64/src/rule.rs [hunk 8]
```diff
@@ -203,42 +318,214 @@ impl Rule<'_> {
...
+
+        impl PartialEq<LazyRuleEval<'_>> for DateTime {
+            fn eq(&self, other: &LazyRuleEval) -> bool {
+                self.partial_cmp(other) == Some(Ordering::Equal)
+            }
+        }
+
+        impl PartialOrd<LazyRuleEval<'_>> for DateTime {
+            fn partial_cmp(&self, rule: &LazyRuleEval) -> Option<Ordering> {
+                let mut rule_day_in_year = rule.rule.day_in_year_and_ny(rule.year).0;
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs]: thought: I'm worried about this getting recomputed multiple times; but hopefully it collapses?

Could perhaps have a lazy field on LazyRuleEval

File: utils/zoneinfo64/src/rule.rs [hunk 9]
```diff
@@ -180,19 +188,126 @@ impl TzRuleDate {
...
+        match self.time_mode {
+            TimeMode::Utc => seconds_of_day,
+            TimeMode::Standard => seconds_of_day - standard_offset_seconds,
+            TimeMode::Wall => {
+                seconds_of_day - (standard_offset_seconds + additional_offset_seconds)
+            }
+        }
+    }
+
+    /// Converts the {transition_time} into a time in the local day, in seconds
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs]: nit: clarify that it is in the post-transition offset.

File: utils/zoneinfo64/src/rule.rs [hunk 10]
```diff
@@ -180,19 +188,126 @@ impl TzRuleDate {
...
+    fn transition_time_to_wall(
+        &self,
+        standard_offset_seconds: i32,
+        additional_offset_seconds: i32,
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs:300]: nit: clarify that this is not `inner.additional_offset_seconds` but rather "additional offset seconds if rule applies, 0 if not"

File: utils/calendrical_calculations/src/chinese_based.rs (hunk 4 above)
Comment (robertbastian) [reply utils/calendrical_calculations/src/chinese_based.rs:39]: this is in a const block

File: utils/zoneinfo64/src/rule.rs (hunk 5 above)
Comment (robertbastian) [reply utils/zoneinfo64/src/rule.rs]: this is in a const block

File: utils/zoneinfo64/src/rule.rs (hunk 8 above)
Comment (robertbastian) [reply utils/zoneinfo64/src/rule.rs]: every `LazyRuleEval` only gets compared once

File: utils/zoneinfo64/src/rule.rs [hunk 11]
```diff
@@ -323,57 +323,40 @@ impl Rule<'_> {
...
-        };
+        let datetime = SecondsInLocalYear(day_in_year as i64 * SECONDS_IN_UTC_DAY + local_time_of_day as i64);
 
         struct LazyRuleEval<'a> {
             year: i32,
             day_before_year: RataDie,
             rule: &'a TzRuleDate,
             standard_offset_seconds: i32,
             additional_offset_seconds: i32,
+            // The correction that is applied after calculating the local time.
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs]: issue: This still doesn't mean anything. What are we correcting here?

We should probably also document what LazyRuleEval _is_

File: utils/zoneinfo64/src/rule.rs [hunk 12]
```diff
@@ -323,57 +323,40 @@ impl Rule<'_> {
             return None;
         }
 
-        struct DateTime {
-            day_in_year: u16,
-            local_time_of_day: i32,
-        }
+        struct SecondsInLocalYear(i64);
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs:328]: thought: Okay, this is preferable. Potentially cite the tested invariant that we never straddle year boundaries.

issue: I'd prefer if this type lived outside of a function. Also LazyRuleEval.

File: utils/calendrical_calculations/src/iso.rs [hunk 13]
```diff
@@ -87,28 +75,36 @@ pub(crate) const fn iso_year_from_fixed(date: RataDie) -> i64 {
...
     } else {
-        year + 1
+        i64_to_i32(year + 1)
     }
 }
 
-fn iso_new_year(year: i32) -> RataDie {
-    fixed_from_iso(year, 1, 1)
+/// Calculates the day before Jan 1 of `year`.
+pub const fn day_before_year(year: i32) -> RataDie {
```
Comment (Manishearth) [inline utils/calendrical_calculations/src/iso.rs:85]: issue: Add some tests for a range of year 2000 ± 50y. I don't want to verify the math for boundary cases

File: utils/zoneinfo64/src/rule.rs (hunk 5 above)
Comment (Manishearth) [reply utils/zoneinfo64/src/rule.rs]: This isn't?

File: utils/zoneinfo64/src/rule.rs [hunk 14]
```diff
@@ -203,42 +314,198 @@ impl Rule<'_> {
...
+            } else if datetime < before_end {
+                Some(PossibleOffset::Single(rule_offset))
+            } else if datetime < after_end {
+                Some(PossibleOffset::Ambiguous(rule_offset, standard_offset))
+            } else {
+                Some(PossibleOffset::Single(standard_offset))
+            }
+        } else {
+            if datetime < before_end {
+                // Here the rule_offset is fine even if year == start_year, as inverted rules seem
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs:441]: suggestion: perhaps pick up the last-transition-smoothly-into-rule invariant test from my PR

File: utils/calendrical_calculations/src/iso.rs (hunk 13 above)
Comment (robertbastian) [reply utils/calendrical_calculations/src/iso.rs:85]: you don't have to, this is preexisting code

File: utils/zoneinfo64/src/rule.rs [hunk 15]
```diff
@@ -203,42 +314,202 @@ impl Rule<'_> {
...
+        let rule_offset = Offset {
+            offset: UtcOffset::from_seconds_unchecked(
+                self.standard_offset_seconds + self.inner.additional_offset_secs,
+            ),
+            rule_applies: true,
+        };
+
+        #[allow(clippy::collapsible_else_if)] // symmetry
+        if !self.inner.is_inverted() {
+            if datetime < before_start {
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs:419]: issue: I noted this when talking about is_inverted but this code has the problem that I need to simultaneously remember which offsets are gap/overlap offsets when the rules are northern or southern/inverted. This is brittle and harder to understand when debugging.

suggestion: you can do the same trick that #6874 does and instead of comparing against start/end, compare against "range" which has been constructed so that it is (first transition...second transition). Then we only need to talk about inverted/regular offsets when constructing that range, and after that we talk abstractly in terms of in/out of range.

Comment (robertbastian) [reply utils/zoneinfo64/src/rule.rs:419]: with the range you have to eagerly evaluate both transitions though

File: utils/zoneinfo64/src/rule.rs [hunk 16]
```diff
@@ -467,44 +468,47 @@ impl Rule<'_> {
...
-            since: self.inner.end.timestamp_for_year(
-                year,
-                day_before_year,
-                self.standard_offset_seconds,
-                self.inner.additional_offset_secs,
-            ),
-            offset: UtcOffset::from_seconds_unchecked(self.standard_offset_seconds),
-            rule_applies: false,
-        };
+        let start = (&self.inner.start, self.standard_offset_seconds, 0, false);
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs]: suggestion: probably worth a struct (preferably not stuck within the function) with a `.transitions_for_year` wrapper method.

File: utils/zoneinfo64/src/rule.rs [hunk 17]
```diff
@@ -471,13 +471,8 @@ impl Rule<'_> {
...
         let day_before_year = iso::day_before_year(year);
 
-        let start = (&self.inner.start, self.standard_offset_seconds, 0, false);
-        let end = (
-            &self.inner.end,
-            self.standard_offset_seconds,
-            self.inner.additional_offset_secs,
-            true,
-        );
+        let start = (&self.inner.start, 0);
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs:485]: okay, this doesn't need a struct anymore.

File: utils/zoneinfo64/src/rule.rs (hunk 15 above)
Comment (Manishearth) [reply utils/zoneinfo64/src/rule.rs:419]: How so? You'd have two ranges; range_prev and range_next, which contain `before_start..before_end` and `after_start..after_end` (the LazyRuleEvals). You have these four values already precomputed, and picking which one to construct is a matter of checking `is_inverted`

The lazy evaluation still works.

Comment (robertbastian) [reply utils/zoneinfo64/src/rule.rs:419]: Your code has a lot more comparisons as it has to handle both the case where the first transition has a gap, as well as a duplicated hour. This wouldn't work well with lazy evaluation.

Comment (Manishearth) [reply utils/zoneinfo64/src/rule.rs:419]: Ah, hm. In that case, can you have more documentation here that explains the math, specifically that gap transitions occur for forward, overlap for back?

And explicitly mention that we expect `additional_offset_secs` to always be positive, with an invariant test for the same. This stops working the moment we get data that has daylight wastings time. Which AIUI ICU4C doesn't wish to do, but we should guard against that. 


Or return an inverted PossibleOffset (gap vs overlap) based on additional_offset_secs.

----

FWIW, telling the difference between the overlap and gap doesn't require a full evaluation; it can be done by comparing the two offsets.

Comment (Manishearth) [reply utils/zoneinfo64/src/rule.rs:419]: happy to make a PR based on this one that attempts this refactor.

File: utils/zoneinfo64/src/rule.rs [hunk 18]
```diff
@@ -417,7 +417,7 @@ impl Rule<'_> {
         };
 
         #[allow(clippy::collapsible_else_if)] // symmetry
-        if !self.inner.is_inverted() {
+        if !self.inner.end_before_start() {
             if datetime < before_start {
                 // Before spring-forward
                 if year == self.start_year as i32 {
```
Comment (Manishearth) [inline utils/zoneinfo64/src/rule.rs:421]: thought: spring-forward fall-back terminology is unfortunate, since in the regions that have "fall-forward spring-back" it is actually still spring-forward fall-back, BUT their spring is in Augustish.

File: utils/zoneinfo64/src/rule.rs (hunk 15 above)
Comment (robertbastian) [reply utils/zoneinfo64/src/rule.rs:419]: > Ah, hm. In that case, can you have more documentation here that explains the math, specifically that gap transitions occur for forward, overlap for back?

Isn't that obvious from spring-forward, fall-back?

> And explicitly mention that we expect additional_offset_secs to always be positive

I already mention this in a comment above this block, but don't test it. 

> Or return an inverted PossibleOffset (gap vs overlap) based on additional_offset_secs.

What?

File: utils/zoneinfo64/src/rule.rs (hunk 18 above)
Comment (robertbastian) [reply utils/zoneinfo64/src/rule.rs:421]: oh right it is

File: utils/calendrical_calculations/src/iso.rs (hunk 13 above)
Comment (Manishearth) [reply utils/calendrical_calculations/src/iso.rs:85]: Ah, right

File: utils/zoneinfo64/src/rule.rs (hunk 15 above)
Comment (Manishearth) [reply utils/zoneinfo64/src/rule.rs:419]: > Isn't that obvious from spring-forward, fall-back?

Not 100%. I've found that when talking about stuff like this there's often a variance problem, e.g. whether you are talking about the transformation to apply to the values in a transformed space to produce a new transformed space, or the transformation to apply to the space's basis.

In _this_ case, fortunately the two most common concepts ("the amount to physically move the clock") and "the delta in the UTC offset" are both of the same variance. This means that it's unlikely for someone to have the _wrong_ impression here, but it is still likely for them to have the right impression but not be sure if it is right.

But this is the same type of confusion that leads to graphics programmers inserting `-` signs until their errors go away.

> What?

I'm suggesting you can handle the additional_offset_secs problem by returning gap instead of overlap and vice versa when additional_offset_secs is negative.

But it's not something I require here.

_That's_ the only reason my code performs extra checks, because it doesn't assume anything about additional_offset_secs.

Anyway, irrelevant.

**Comment by robertbastian** [comment]
I think I should clarify the invariants a bit more, but if you want to merge and keep working on it go for it.
